akka-operator/templates/040-deployment.yaml

```diff
@@ -1,3 +1,86 @@
+{{- if eq .Values.cloudProvider.name "gcp" }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: $name-ubbagent-config
+data:
+  config.yaml: |
+    # The identity section contains authentication information used
+    # by the agent.
+    identities:
+    - name: gcp
+      gcp:
+        # This parameter accepts a base64-encoded JSON service
+        # account key. The value comes from the reporting secret.
+        encodedServiceAccountKey: $AGENT_ENCODED_KEY
+
+    # The metrics section defines the metric that will be reported.
+    # Metric names should match verbatim the identifiers created
+    # during pricing setup.
+    metrics:
+    - name: requests
+      type: int
+
+      # The endpoints section of a metric defines which endpoints the
+      # metric data is sent to.
+      endpoints:
+      - name: on_disk
+      - name: servicecontrol
+
+      # The aggregation section indicates that reports that the agent
+      # receives for this metric should be aggregated for a specified
+      # period of time prior to being sent to the reporting endpoint.
+      aggregation:
+        bufferSeconds: 60
+
+    - name: instance_time
+      type: int
+      endpoints:
+      - name: on_disk
+      - name: servicecontrol
+
+      # The passthrough marker indicates that no aggregation should
+      # occur for this metric. Reports received are immediately sent
+      # to the reporting endpoint. We use passthrough for the
+      # instance_time metric since reports are generated
+      # automatically by a heartbeat source defined in a later
+      # section.
+      passthrough: {}
+
+    # The endpoints section defines where metering data is ultimately
+    # sent. Currently supported endpoints include:
+    # * disk - some directory on the local filesystem
+    # * servicecontrol - Google Service Control
+    endpoints:
+    - name: on_disk
+      # The disk endpoint is useful for debugging, but its inclusion
+      # is not necessary in a production deployment.
+      disk:
+        reportDir: /var/lib/ubbagent/reports
+        expireSeconds: 3600
+    - name: servicecontrol
+      servicecontrol:
+        identity: gcp
+        # The service name is unique to your application and will be
+        # provided during onboarding.
+        serviceName: my-application.mp-my-company.appspot.com
+        consumerId: $AGENT_CONSUMER_ID  # From the reporting secret.
+      
+
+    # The sources section lists metric data sources run by the agent
+    # itself. The currently-supported source is 'heartbeat', which
+    # sends a defined value to a metric at a defined interval. In
+    # this example, the heartbeat sends a 60-second value through the
+    # "instance_time" metric every minute.
+    sources:
+    - name: instance_time_heartbeat
+      heartbeat:
+        metric: instance_time
+        intervalSeconds: 60
+        value:
+          int64Value: 60
+{{- end }}
+---
```

Comment:
this is the config map, still need to add the side car?

Reply:
The metering agent's configuration is included as a ConfigMap value. When deployed as a sidecar, the agent's configuration can be parameterized through the use of environment variables, which is the mechanism used for passing values from the reporting secret. More details at https://github.com/GoogleCloudPlatform/marketplace-k8s-app-tools/blob/64181befcb4d3a5417e84d4f59fea82b016988ab/docs/billing-integration.md#agent-deployment-and-configuration

Reply:
But you still need to add the sidecar right? Not just the config map

```
- name: ubbagent
        image: $imageUbbagent
        env:
        - name: AGENT_CONFIG_FILE
          value: "/etc/ubbagent/config.yaml"
        - name: AGENT_LOCAL_PORT
          value: "4567"
        - name: AGENT_ENCODED_KEY
          valueFrom:
            secretKeyRef:
              name: $reportingSecret
              key: reporting-key
        - name: AGENT_CONSUMER_ID
          valueFrom:
            secretKeyRef:
              name: $reportingSecret
              key: consumer-id
        volumeMounts:
        - name: ubbagent-config
          mountPath: /etc/ubbagent
```

Reply:
@chbatey Yes, I will add the sidecar container too. But this will again have to be conditional depending on whether we are doing gcp or not. In such cases where we target multiple deployment platforms, is it idiomatic to have a single yaml for deployment with such if/else conditionals. It's becoming a bit messy I think. Or is it more idiomatic to have some sort of separation of yamls that we aggregate during the build process depending on the target deployment platform ? I don't have much k8s fu - looking for suggestions :-)

Reply:
added the sidecar now inline with conditionals